KIP-74: Add fetch response size limit and implement round-robin on client side

nepal opened the pull request:
This PR is implementation of [KIP-74](https://cwiki.apache.org/confluence/display/KAFKA/KIP-74%3A+Add+Fetch+Response+Size+Limit+in+Bytes) which is originally motivated by [KAFKA-2063](https://issues.apache.org/jira/browse/KAFKA-2063).

Your comments are greatly appreciated.

Commit:
KAFKA-2063: Add possibility to bound fetch response size

ijuma commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -278,6 +279,7 @@ object KafkaConfig {
   val ReplicaFetchMaxBytesProp = "replica.fetch.max.bytes"
```
I wonder if we should deprecate this in favour of a `replica.fetch.max.partition.bytes`. It seems quite confusing that this one applies to a single partition when every other `replica.fetch.*` setting applies to the request. Maybe something for later.

benstopford replied:
gets my vote

hachikuji replied:
It's tempting to do this now while we're still in the KIP window. We'll be less inclined to do it in the future if it requires a new KIP.

ijuma replied:
What you say is true, but I think it's too late to add this to the KIP as we don't have a great story for deprecating properties.

ijuma commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -476,6 +478,7 @@ object KafkaConfig {
   val ReplicaFetchWaitMaxMsDoc = "max wait time for each fetcher request issued by follower replicas. This value should always be less than the " +
   "replica.lag.time.max.ms at all times to prevent frequent shrinking of ISR for low throughput topics"
   val ReplicaFetchMinBytesDoc = "Minimum bytes expected for each fetch response. If not enough bytes, wait up to replicaMaxWaitTimeMs"
+  val ReplicaFetchResponseMaxBytesDoc = "Maximum bytes expected for entire fetch respones."
```
`respones` should be `response`. Also, it's probably a good idea to clarify the doc for `ReplicaFetchMaxBytesDoc` as it's a bit vague.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -113,6 +113,13 @@
+     * <code>fetch.max.bytes</code>
+     */
+    public static final String FETCH_MAX_BYTES_CONFIG = "fetch.max.bytes";
+    private static final String FETCH_MAX_BYTES_DOC = "The maximum amount of data the server should return for a fetch request.";
```
Maybe worth mentioning that this is not an absolute maximum and if a single message is larger than the maximum, it will still be returned.

nepal replied:
OK

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +63,67 @@ public PartitionData(long offset, int maxBytes) {
+            struct.set(RESPONSE_MAX_BYTES_KEY_NAME, maxBytes);
         List<Struct> topicArray = new ArrayList<Struct>();
-        for (Map.Entry<String, Map<Integer, PartitionData>> topicEntry : topicsData.entrySet()) {
+        for (TopicAndPartitionData topicEntry : topicsData) {
```
The fact that the partitions are grouped by topic means that we don't have full flexibility when implementing the round-robin fetch. Have you thought about how this will work exactly?

nepal replied:
method groupByTopicOrdered() will group partitions by topic and preserve order.
So, for example,  the following sequence of TopicPartition's {(t1, p1), (t1, p2), (t2, p3), (t1, p4)} will be grouped as
{(t1, {p1,p2}), (t2, {p3}), (t1, {p4})
Here () - is a tuple, {} - is a sequence.

So I see no problems with round-robin fetch here.

ijuma replied:
OK, I see, the topic can appear multiple lines in the list (so not a traditional `groupBy`). That works and it can be more efficient (if each topic group has multiple partitions) than an array of TopicPartitions (which is conceptually simpler, in my opinion). Let's see what @junrao says.

benstopford replied:
I wondered the same. Adding a test could be a good way to highlight this.

hachikuji replied:
It's definitely a little weird to let the fetch API have multiple instances of the same topic, but that seems like a reasonable way to preserve ordering by partition which keeping fetch requests small. Repeating the same topic multiple times could be a problem when you have many partitions. As long as the partitions being fetched are ordered by topic (modulo the number of partitions), this ensures that only one topic is listed more than once.

cc @edenhill

hachikuji replied:
Let me clarify what I meant about maintaining the order of the fetched partitions by topic. If we have topics a, b and c, each with two partitions, we might initially send:

a0, a1, b0, b1, c0, c1

If data returns for a0, a1 and b0, then we'd reorder to:

b1, c0, c1, a0, a1, b0

So only one of the partitions wraps around and would be listed twice in the request. The downside of ordering this way is that it could result in delays fetching from certain topics if one or several other topics have many partitions and are particularly busy.

becketqin replied:
@hachikuji 's solution sounds reasonable to me. One minor thing is that we may want to keep the last partition that has some data returned at the beginning if there are more than one partition returned data. Otherwise theoretically what could happen (although unlikely) is a consumer subscribed to a topic and the larger partitions always consume less data than the earlier topics due to the cap of response size. For example, if a consumer subscribe to a0, a1, a2 and specified the max response size to be 5 MB, and max bytes per partition to be 2 MB, what could happen is a0, a1, a2 always return 2 MB, 2 MB and 1 MB respectively. So a2 might be always lagging.

ijuma commented on core/src/main/scala/kafka/api/FetchRequest.scala:
```diff
@@ -56,7 +58,7 @@ object FetchRequest {
       })
     })
-    FetchRequest(versionId, correlationId, clientId, replicaId, maxWait, minBytes, Map(pairs:_*))
+    FetchRequest(versionId, correlationId, clientId, replicaId, maxWait, minBytes, maxBytes, ListMap(pairs:_*))
```
`ListMap` is not a good collection as it performs really badly unless its size is small (`get` is O(n) for example).

ijuma commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1021,6 +1026,7 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean) extends Abstra
     require(replicaFetchWaitMaxMs <= replicaSocketTimeoutMs, "replica.socket.timeout.ms should always be at least replica.fetch.wait.max.ms" +
       " to prevent unnecessary socket timeouts")
     require(replicaFetchMaxBytes >= messageMaxBytes, "replica.fetch.max.bytes should be equal or greater than message.max.bytes")
+    require(replicaFetchResponseMaxBytes >= 0, "replica.fetch.response.max.bytes should be equal or greater than 0")
```
You should be able to use a validator like you did for the consumer config.

ijuma commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -538,13 +541,22 @@ class ReplicaManager(val config: KafkaConfig,
               FetchDataInfo(LogOffsetMetadata.UnknownOffsetMetadata, MessageSet.Empty)
           }
 
+          if (limitBytes != Int.MaxValue)
```
Why do we need this check?

nepal replied:
According to KIP, if limitBytes is equal to Int.MaxValue the fetch request should be processed exactly like pre v.3 request. Maybe we should setup a special flag for this:
val isOldRequest = (limitBytes == Int.MaxValue)
and remove this check

ijuma replied:
Yes, I understood that compatibility was the intent, but I was trying to figure out how it would behave differently without the check. You're thinking of cases where the response size would be larger than 2^31?

ijuma commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -538,13 +541,22 @@ class ReplicaManager(val config: KafkaConfig,
           }
 
+          if (limitBytes != Int.MaxValue)
+            limitBytes -= Math.min(limitBytes, logReadInfo.messageSet.sizeInBytes)
```
Maybe it's just me, but I find the following clearer:

``` scala
limitBytes = math.max(0, limitBytes - logReadInfo.messageSet.sizeInBytes)
```

```
```

benstopford replied:
seems a bit clearer to me too.

ijuma commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -538,13 +541,22 @@ class ReplicaManager(val config: KafkaConfig,
+                if (Topic.isInternal(topic) && !readOnlyCommitted)
+                  Math.max(fetchSize, log.config.maxMessageSize)
+                else if (limitBytes > 0)
+                  fetchSize
```
Why are we using `fetchSize` without considering `limitBytes` here?

nepal replied:
As I mentioned, there should be something like that
adjustedFetchSize = max(sizeOfNextMessageInPartition, limitBytes)

ijuma commented:
Thanks for the PR @nepal. It looks to me like it's going in the right direction. Left some comments/questions and looking forward to the missing bits (when do you think you'll have them?). By the way, I would add one point to your to-do list: tests.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +63,67 @@ public PartitionData(long offset, int maxBytes) {
+        this(replicaId, maxWait, minBytes, fetchData, ProtoUtils.latestVersion(ApiKeys.FETCH.id), maxBytes);
+    }
+
+    public FetchRequest(int replicaId, int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData, int version, int maxBytes) {
```
I suggest making the `version` the first parameter.

junrao replied:
Since the ordering is relevant here, would it be better to use LinkedHashMap in the signature?

nepal replied:
Not sure whether it is good idea to accept only LinkedHashMap in constructor - it is used for all versions of FetchRequest. Maybe we should just add comment that FetchRequest will preserve the ordering of fetchData.

ijuma replied:
Using LinkedHashMap for all versions should be fine, right? The previous versions don't care about ordering, so either way is fine. Or am I missing something?

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -100,13 +100,13 @@ public FetchResponse(Map<TopicPartition, PartitionData> responseData, int thrott
         initCommonFields(responseData);
         struct.set(THROTTLE_TIME_KEY_NAME, throttleTime);
-        this.responseData = responseData;
+        this.responseData = new LinkedHashMap<TopicPartition, PartitionData>(responseData);
```
Could we change the comment that this is the constructor for version 1,2 and 3?

junrao replied:
It would be clearer if we can have a constructor for v3 using LinkedHashMap. This may not be possible though. So, it would be useful to at least add a comment that for v3, we expect the caller to pass in a LinkedHashMap since the ordering is relevant.

junrao commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -54,7 +54,9 @@ object ApiVersion {
-    "0.10.1-IV0" -> KAFKA_0_10_1_IV0
+    "0.10.1-IV0" -> KAFKA_0_10_1_IV0,
+    // 0.10.1-IV1 is introduced for KIP-74(fetch response size limit).
+    "0.10.1-IV1" -> KAFKA_0_10_1_IV1
```
Could we add the following entry at the end?

```
 "0.10.1" -> KAFKA_0_10_1_IV1
```

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -507,15 +508,17 @@ class ReplicaManager(val config: KafkaConfig,
       val partitionDataAndOffsetInfo =
         try {
-          trace("Fetching log segment for topic %s, partition %d, offset %d, size %d".format(topic, partition, offset, fetchSize))
+          trace("Fetching log segment for topic %s, partition %d, offset %d, size %d, limit %d".format(topic, partition, offset, fetchSize, limitBytes))
```
It's not very clear what's the difference btw size and limit from the log message. Could we change it to partition size and response limit?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -538,13 +541,22 @@ class ReplicaManager(val config: KafkaConfig,
             case Some(log) =>
-              val adjustedFetchSize = if (Topic.isInternal(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
+              val adjustedFetchSize =
+                if (Topic.isInternal(topic) && !readOnlyCommitted)
```
If we always return the first message independent of its size, there is no need to specialize for the internal topic any more. Could we remove this check?

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -432,6 +432,23 @@
+                                                                       "Maximum bytes to accumulate in the response."),
+                                                             new Field("topics",
+                                                                       new ArrayOf(FETCH_REQUEST_TOPIC_V0),
+                                                                       "Topics to fetch."));
```
Could we add a comment that the ordering of those topic/partitions are now relevant?

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +63,67 @@ public PartitionData(long offset, int maxBytes) {
+        }
+
+        public static final List<TopicAndPartitionData> groupByTopicOrdered(Map<TopicPartition, PartitionData> fetchData) {
+            List<TopicAndPartitionData> topics = new ArrayList<TopicAndPartitionData>();
```
ArrayList<TopicAndPartitionData> can be ArrayList<>. Ditto for LinkedHashMap in line 72. There are a few other places like that.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +63,67 @@ public PartitionData(long offset, int maxBytes) {
+            List<TopicAndPartitionData> topics = new ArrayList<TopicAndPartitionData>();
+            for (Map.Entry<TopicPartition, PartitionData> topicEntry : fetchData.entrySet()) {
+                String topic = topicEntry.getKey().topic();
+                int parition = topicEntry.getKey().partition();
```
typo parition

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +63,67 @@ public PartitionData(long offset, int maxBytes) {
+                String topic = topicEntry.getKey().topic();
+                int parition = topicEntry.getKey().partition();
+                PartitionData partitionData = topicEntry.getValue();
+                if (topics.isEmpty() || topics.get(topics.size() - 1).topic != topic) {
```
No need for {} for single-line statement.

junrao commented:
@nepal : Thanks for the patch. A couple of other comments.
1. Could you change the upgrade doc about the protocol change since we now require setting inter.broker.protocol for rolling upgrade to 0.10.1?
2. In seems that we need to change FetchResponse.scala to take into consideration of the ordering of the partitions.

ijuma commented:
@nepal any idea when you will update the PR with the missing bits? Let us know if you need any help making progress on this. Thanks!

nepal commented:
I am currently working on missing bits. I think it will take three days. Maybe we can do some intermediate version of this PR which can be accepted if I am blocking someone?

ijuma commented:
Thanks for the update @nepal, 3 days is fine. Probably easier to get the one PR in since it's not too complicated.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -113,6 +113,13 @@
+     * <code>fetch.max.bytes</code>
+     */
+    public static final String FETCH_MAX_BYTES_CONFIG = "fetch.max.bytes";
+    private static final String FETCH_MAX_BYTES_DOC = "The maximum amount of data the server should return for a fetch request. This is not an absolute maximum - if there is a single message which is larger than fetch.max.bytes, it will still be returned.";
```
Do you think it's worth mentioning here that the consumer may send multiple parallel fetches and that this setting only applies to each fetch individually? In a future KIP, it might make sense to add a `max.in.flight.fetches` setting so that users have better control over memory. I went ahead and opened a JIRA to collect comments on this: https://issues.apache.org/jira/browse/KAFKA-4133.

becketqin replied:
Another approach might be KIP-72 which provides a memory pool to control the memory footprint.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -33,6 +32,7 @@
     private static final String REPLICA_ID_KEY_NAME = "replica_id";
     private static final String MAX_WAIT_KEY_NAME = "max_wait_time";
     private static final String MIN_BYTES_KEY_NAME = "min_bytes";
+    private static final String RESPONSE_MAX_BYTES_KEY_NAME = "max_bytes";
```
nit: should this be `MAX_BYTES_KEY_NAME` to be consistent with `MIN_BYTES_KEY_NAME`?

hachikuji replied:
Oh, I see we already had one defined below. Maybe we could just share the constant or change the other definition to `PARTITION_MAX_BYTES_KEY_NAME`?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -44,10 +44,14 @@
     private static final String MAX_BYTES_KEY_NAME = "max_bytes";
 
+    // default values for current version
+    public static final int DEFAULT_RESPONSE_MAX_BYTES = Integer.MAX_VALUE;
```
I'm a little confused. Is this for older versions? If it's for the current version, shouldn't it match `FETCH_MAX_WAIT_MS_CONFIG` in `ConsumerConfig`?

becketqin replied:
Does "current version" mean "v0, v1 and v2"?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +63,66 @@ public PartitionData(long offset, int maxBytes) {
     /**
      * Create a non-replica fetch request
      */
     public FetchRequest(int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData) {
```
Since the order is now significant, maybe we should consider changing these constructors so that they either accept `LinkedHashMap` or `List<PartitionData>`? If there are compatibility concerns, at least we should do this for _new_ constructors.

becketqin replied:
Should we just keep this constructor return the v2 request and maybe deprecate it? It seems that `maxBytes` should be a mandatory field for FetchRequest v3.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +63,66 @@ public PartitionData(long offset, int maxBytes) {
+            this.partitions = new LinkedHashMap<>();
+        }
+
+        public static final List<TopicAndPartitionData> groupByTopicOrdered(Map<TopicPartition, PartitionData> fetchData) {
```
nit: does `final` make sense?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -121,7 +167,7 @@ public FetchRequest(Struct struct) {
     @Override
     public AbstractRequestResponse getErrorResponse(int versionId, Throwable e) {
-        Map<TopicPartition, FetchResponse.PartitionData> responseData = new HashMap<TopicPartition, FetchResponse.PartitionData>();
+        Map<TopicPartition, FetchResponse.PartitionData> responseData = new LinkedHashMap<TopicPartition, FetchResponse.PartitionData>();
```
nit: there are a couple cases like this where the generic types can be omitted.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -65,7 +65,7 @@
     public static final ByteBuffer EMPTY_RECORD_SET = ByteBuffer.allocate(0);
 
-    private final Map<TopicPartition, PartitionData> responseData;
+    private final LinkedHashMap<TopicPartition, PartitionData> responseData;
```
Is there a specific reason to preserve the order? I'm guessing we'll end up looking up by key anyway.

junrao replied:
It seems that the ordering in the response is important in figuring out the partition ordering in the next fetch request?

hachikuji replied:
I assumed we would use the ordering from the request, but it's probably safer to preserve the ordering in the response as well.

ijuma commented:
@nepal the release plan for 0.10.1.0 has been posted and the feature freeze is on the 19th September. So we are on a tight schedule. :) Have you had a chance to work on the missing bits and do you have an ETA? Let us know if we can help move this forward. Thanks!

Commit:
LOGBROKER-2063: Change interfaces of FetchRequest/Response to Vector instead of Map

nepal commented:
I've pushed interface change for FetchRequest/Response. The remaining bits are -
1) tests
2) round robin logic (in AbstractFetcherThread?).

@ijuma  - not sure if I can finish this until feature freeze, so maybe it is a good time for you to step in.

Tim-Brooks committed:
KAFKA-2311; Make KafkaConsumer's ensureNotClosed method thread-safe

Here is the patch on github ijuma.

Acquiring the consumer lock (the single thread access controls) requires that the consumer be open. I changed the closed variable to be volatile so that another thread's writes will visible to the reading thread.

Additionally, there was an additional check if the consumer was closed after the lock was acquired. This check is no longer necessary.

This is my original work and I license it to the project under the project's open source license.

Author: Tim Brooks <tim@uncontended.net>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #1637 from tbrooks8/KAFKA-2311

ijuma committed:
Refactor constructors for a.k.c.r.FetchRequest and introduce factory methods for the replica case

Use diamond in a couple of places.

ijuma committed:
Add missing cases in `FetchRequest.getErrorResponse` and equality bug

ijuma committed:
Fix reference to assignFromUser in javadoc

ijuma committed:
Use `TopicPartition` instead of `TopicAndPartition` in *FetcherThread classes

This will make it easier to reuse a shared Java class for request round-robin.

lindong28 committed:
KAFKA-4158; Reset quota to default value if quota override is deleted

Author: Dong Lin <lindong28@gmail.com>

Reviewers: Joel Koshy <jjkoshy.w@gmail.com>, Jiangjie Qin <becket.qin@gmail.com>

Closes #1851 from lindong28/KAFKA-4158

ijuma committed:
Implement round-robin for `Fetcher` and `*FetcherThread` classes

ijuma committed:
MINOR: Give a name to the coordinator heartbeat thread

Followed the same naming pattern as the producer sender thread.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson

Closes #1854 from ijuma/heartbeat-thread-name

Commit:
KAFKA-4162: Fixed typo "rebalance"

Author: David Chen <mvjome@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1853 from mvj3/KAFKA-4162

ijuma committed:
Flesh out tests in `RequestResponseTest`

ijuma committed:
Flesh out config descriptions in `ConsumerConfig`

hachikuji committed:
KAFKA-4172; Ensure fetch responses contain the requested partitions

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1857 from hachikuji/KAFKA-4172

ijuma committed:
Various compatibility and ordering changes to fetch request/response classes

ijuma committed:
Introduce `minOneMessage` parameter to `Log.read` with a `minOneMessage`

ijuma committed:
Use `Seq` instead of `Map` in `ReplicaManager` and improve handling of messages larger than fetch size

We now always return the first message of the first partition (if there is one) irrespective of its size.

ijuma committed:
Fix comment in `DEFAULT_RESPONSE_MAX_BYTES`

ijuma committed:
Mention broker configs for limiting message size

ijuma committed:
Allow `message.max.bytes` to be smaller than `replica.fetch.max.bytes`

ijuma committed:
Add validator for `ReplicaFetchMaxBytesProp` and reduce priority to medium

ijuma committed:
Rename `groupByTopicOrdered` to `batchByTopic` to be consistent

ijuma committed:
Combine "max_bytes` key name constants

ijuma committed:
Remove `final` from static method

ijuma committed:
Use `batchByTopic` in `FetchResponse` too

nepal committed:
Merge pull request #1 from ijuma/kafka-2063-bound-fetch-response-size

KAFKA-2063: Bound fetch response size (KIP-74)

ijuma committed:
Add tests to verify that we only return a large message if it's the first message in the first partition

ijuma committed:
Rename `partitionMap` to `fetchBuilder` in `AbstractFetcherThread`

ijuma commented on clients/src/main/java/org/apache/kafka/common/internals/FetchBuilder.java:
```diff
@@ -0,0 +1,150 @@
+public class FetchBuilder<S> {
+
+    private final List<PartitionState<S>> list = new ArrayList<>();
+    private final Map<TopicPartition, S> map = new HashMap<>();
```
We don't need both `list` and `map` here as we only ever move items to the end. So I will remove the `list` after adding some tests for this class.

ijuma replied:
This is done now.

ijuma committed:
Add test for `FetchBuilder` and document it

ijuma committed:
Remove `list` from `FetchBuilder` and change `map` to be a `LinkedHashMap`

ijuma committed:
Fix import checkstyle error

nepal committed:
Merge pull request #2 from ijuma/kafka-2063-bound-fetch-response-size

KAFKA-2063: Bound fetch response size (KIP-74)

ijuma commented on clients/src/main/java/org/apache/kafka/common/internals/FetchBuilder.java:
```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    public static class PartitionState<S> {
```
@hachikuji, we could potentially remove this too and just return the `LinkedHashMap`.

hachikuji replied:
No strong preference on my side, but the encapsulation feels a little stronger using `PartitionState` and it also means one less generic type in the usage.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -556,6 +575,7 @@ private void handleListOffsetResponse(TopicPartition topicPartition,
         int bytes = 0;
         int recordsCount = 0;
         PartitionRecords<K, V> parsedRecords = null;
+        boolean emptyResponse = false;
```
It's a little weird that errors are treated as non-empty responses. Maybe we can avoid the overloading by explicitly checking the error code in the `if` check before the `movePartitionsToEnd` call.

It's probably also helpful to add a comment explaining the justification for moving the partition to the back when an error is encountered.

ijuma replied:
That's fair, maybe I don't even need the `emptyResponse` boolean in that case.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/internals/FetchBuilder.java:
```diff
@@ -0,0 +1,170 @@
+ */
+public class FetchBuilder<S> {
+
+    private final LinkedHashMap<TopicPartition, S> map = new LinkedHashMap<>();
```
Initially, when `set(partitionState)` is called, the partitions are ordered by topic. Over time, the order becomes more confusing since the partitions are fetched and rotated in different orders. As long as partitions stay assigned to the same brokers, the relative order should be preserved for each broker by topic. But as partitions move, it seems like we'll lose that relative ordering. I'm not sure how big of a concern this is since the ordering is just an optimization, but maybe we need to occasionally reset the state to avoid it becoming too disorganized. Either that, or we should do the round-robin per-broker.

ijuma replied:
It's a good question. My current feeling is that it's not worth the additional complexity, but I could be convinced otherwise. It's also something that we can add later if it proves to be required.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/internals/FetchBuilder.java:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    private void update(Map<TopicPartition, S> partitionToState) {
+        LinkedHashMap<String, List<TopicPartition>> topicToPartitions = new LinkedHashMap<>();
```
Minor: could this be a normal `HashMap`? Seems like we just use this for grouping the partitions and the order is not significant.

ijuma replied:
I specified that we respect the order of `partitionToState`. Makes it a bit easier to write tests and seemed a little nicer.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +64,79 @@ public PartitionData(long offset, int maxBytes) {
+    @Deprecated
     public FetchRequest(int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData) {
-        this(CONSUMER_REPLICA_ID, maxWait, minBytes, fetchData);
+        // Any of 0, 1 or 2 would do here
```
nit: maybe mention here that the schemas for these versions are identical?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -96,15 +151,20 @@ public FetchRequest(int replicaId, int maxWait, int minBytes, Map<TopicPartition
         this.minBytes = minBytes;
-        this.fetchData = fetchData;
+        this.maxBytes = maxBytes;
+        this.fetchData = new LinkedHashMap<>(fetchData);
```
We don't have a similar defensive copy in `FetchResponse`. Is that intentional?

ijuma replied:
I didn't mean to have that, it is a relic from when the code was taking a plain `Map`. So I am inclined to remove this copy unless you think we should have it?

hachikuji replied:
Yeah, seems unnecessary.

ijuma committed:
Clean-ups in `FetchRequest`

ijuma committed:
Remove `emptyResponse` boolean in `Fetcher` and add clarifying comment

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2063-bound-fetch-response-size

* apache/trunk:
  HOTFIX: fix KafkaStreams SmokeTest
  KAFKA-4160: Ensure rebalance listener not called with coordinator lock
  KAFKA-4172; Ensure fetch responses contain the requested partitions
  KAFKA-4162: Fixed typo "rebalance"
  MINOR: Give a name to the coordinator heartbeat thread
  KAFKA-4158; Reset quota to default value if quota override is deleted
  KAFKA-2311; Make KafkaConsumer's ensureNotClosed method thread-safe

hachikuji commented on core/src/main/scala/kafka/log/FileMessageSet.scala:
```diff
@@ -128,11 +128,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
   /**
    * Search forward for the file position of the last offset that is greater than or equal to the target offset
-   * and return its physical position. If no such offsets are found, return null.
+   * and return its physical position and the size of the message. If no such offsets are found, return null.
```
Perhaps "... return its physical position and the size of the message **at that offset**"?

hachikuji replied:
Might also be helpful to emphasize that the returned size includes the log overhead.

hachikuji commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -135,33 +136,40 @@ class LogSegment(val log: FileMessageSet,
+    if (startOffsetAndSize == null)
       return null
 
+    val (startPosition, messageSize) = startOffsetAndSize
```
Instead of `messageSize`, maybe this should be `messageSetSize` since it includes the log overhead?

ijuma committed:
Add `testReadWithMinMessage` to `LogTest`

hachikuji commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -135,33 +136,40 @@ class LogSegment(val log: FileMessageSet,
+    if (maxSize == 0 && !minOneMessage)
       return FetchDataInfo(offsetMetadata, MessageSet.Empty)
 
+    val maxLength =
```
The generic naming makes this a little difficult to follow (at a glance, what's the difference between `maxSize` and `maxLength`?). Perhaps `maxSize` could be `requestedMaxSize` and `maxLength` could be `maxSize`?

ijuma replied:
I would prefer not to change the method parameter as that affects callers who use named arguments. `adjustedMaxSize` instead of `maxLength`?

ijuma committed:
Minor clean-ups in `searchOffsetWithSize` and `LogSegment.read`

As per Jason's suggestions.

hachikuji commented on core/src/main/scala/kafka/server/AbstractFetcherManager.scala:
```diff
@@ -85,8 +86,9 @@ abstract class AbstractFetcherManager(protected val name: String, clientId: Stri
-        fetcherThreadMap(brokerAndFetcherId).addPartitions(partitionAndOffsets.map { case (topicAndPartition, brokerAndInitOffset) =>
-          topicAndPartition -> brokerAndInitOffset.initOffset
+        fetcherThreadMap(brokerAndFetcherId).addPartitions(partitionAndOffsets.map {
+          case (TopicAndPartition(t, p), brokerAndInitOffset) =>
```
minor: seems like it would just be a little more work to use `TopicPartition` from the two callers of `addFetcherForPartitions`.

ijuma replied:
Done locally.

hachikuji commented on core/src/main/scala/kafka/server/AbstractFetcherManager.scala:
```diff
@@ -98,7 +100,7 @@ abstract class AbstractFetcherManager(protected val name: String, clientId: Stri
     mapLock synchronized {
       for ((key, fetcher) <- fetcherThreadMap) {
-        fetcher.removePartitions(partitions)
+        fetcher.removePartitions(partitions.map { case TopicAndPartition(t, p) => new TopicPartition(t, p) })
```
Ditto here. It actually looks like it saves an unnecessary conversion to `TopicAndPartition`.

ijuma replied:
Done locally.

ijuma committed:
Replace usages of `TopicAndPartition` with `TopicPartition` in `*FetcherManager` classes

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1025,7 +1036,6 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean) extends Abstra
     require(logCleanerDedupeBufferSize / logCleanerThreads > 1024 * 1024, "log.cleaner.dedupe.buffer.size must be at least 1MB per cleaner thread.")
     require(replicaFetchWaitMaxMs <= replicaSocketTimeoutMs, "replica.socket.timeout.ms should always be at least replica.fetch.wait.max.ms" +
       " to prevent unnecessary socket timeouts")
-    require(replicaFetchMaxBytes >= messageMaxBytes, "replica.fetch.max.bytes should be equal or greater than message.max.bytes")
```
Would it ever make sense for users to set the replica fetch bytes smaller than the max message size (even if it's technically now permitted)?

ijuma replied:
I think this may make sense for cases where large messages are rare but can occur. For those cases, you may want to keep the default `replica.fetch.max.bytes` while allowing for larger messages. What do you think? It would also be good to get @junrao's input on this one.

nepal replied:
As for me, I think it makes perfect sense - without this check we can keep consumer memory footprint low while it can still be able to fetch (rare) big messages.

junrao commented on core/src/main/scala/kafka/api/FetchRequest.scala:
```diff
@@ -23,19 +23,25 @@ import kafka.common.TopicAndPartition
 import java.util.concurrent.atomic.AtomicInteger
 import java.nio.ByteBuffer
+
+import org.apache.kafka.common.TopicPartition
```
unused import

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -517,11 +517,12 @@ class Log(val dir: File,
    * @return The fetch data information including fetch starting offset metadata and messages read.
    */
-  def read(startOffset: Long, maxLength: Int, maxOffset: Option[Long] = None): FetchDataInfo = {
+  def read(startOffset: Long, maxLength: Int, maxOffset: Option[Long] = None, minOneMessage: Boolean = false): FetchDataInfo = {
```
It seems that the caller in GroupMetadataManager should also set minOneMessage to true to avoid getting stuck on a large message.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -457,43 +457,50 @@ class ReplicaManager(val config: KafkaConfig,
-        (topicAndPartition, FetchPartitionStatus(result.info.fetchOffsetMetadata, fetchInfo.get(topicAndPartition).get))
+        val fetchInfo = fetchInfos.collectFirst {
+          case (tp, v) if tp == topicAndPartition => v
+        }.getOrElse(sys.error(s"Partition $topicAndPartition not found in fetchInfos"))
```
Hmm, not sure why we want to do that. It seems that every partition should always be in logReadResults.

ijuma replied:
Generally, it's good practice to use `getOrElse` instead of `get` so that a helpful error message is shown if an invariant is broken. There are and have been places in Kafka where we get a NoSuchElementException when someone incorrectly uses `get`. As you said, it should never happen unless there's a bug.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -503,70 +510,82 @@ class ReplicaManager(val config: KafkaConfig,
+    var limitBytes = fetchMaxBytes
+    val result = new mutable.ArrayBuffer[(TopicAndPartition, LogReadResult)]
+    readPartitionInfo.foreach { case (tp, fetchInfo) =>
+      val minOneMessage = tp == readPartitionInfo.head._1
```
I think we should only turn off minOneMessage after we get the first non-empty partition in readResult.

junrao replied:
If we make that change, we will want to change the docs in KafkaConfig and ConsumerConfig accordingly.

ijuma replied:
I think you're right, I made that change.

junrao commented on clients/src/main/java/org/apache/kafka/common/internals/FetchBuilder.java:
```diff
@@ -0,0 +1,170 @@
+    public List<PartitionState<S>> partitionStates() {
+        List<PartitionState<S>> result = new ArrayList<>();
+        for (Map.Entry<TopicPartition, S> entry : map.entrySet()) {
+            result.add(new PartitionState<S>(entry.getKey(), entry.getValue()));
```
<S> can be <>.

ijuma committed:
Rename `FetchBuilder` to `PartitionStates`

ijuma committed:
Fixed ConcurrentModificationException by returning a copy in `partitionSet`

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2063-bound-fetch-response-size

* apache/trunk:
  KAFKA-1464; Add a throttling option to the Kafka replication
  KAFKA-4131; Multiple Regex KStream-Consumers cause Null pointer exception

ijuma committed:
Improve logging in `ReplicaManager.readFromLocalLog`

ijuma committed:
Improve behaviour when `maxSize` is non-zero but smaller than the message size

It is now consistent with maxSize=0.

ijuma committed:
Add `testLowMaxFetchSizeForRequestAndPartition`

Include a few test clean-ups

ijuma committed:
Update `Fetcher` not to try and detect `RecordTooLargeException`s

The behaviour in brokers changed so it should never happen.

nepal committed:
Merge pull request #3 from ijuma/kafka-2063-bound-fetch-response-size

KAFKA-2063: Address Jason's comments, add tests, fix bugs, merge trunk

ijuma committed:
Accept large messages from the first non-empty partition (instead of the first partition)

As suggested by Jun

ijuma committed:
Use `minOneMessage = true` when reading the log from `GroupMetadataManager`

ijuma committed:
Remove unused import.

ijuma committed:
Remove `warnIfMessageOversized` from `ReplicaFetcherThread`

ijuma committed:
Remove message size warnings from `TopicCommand` and `ConfigCommand`

ijuma committed:
Add test to verify that oversized messages produced with ack=all succeed

Oversized in the context of replication.

nepal committed:
Merge pull request #4 from ijuma/kafka-2063-bound-fetch-response-size

KAFKA-2063: Address Jun's comments, remove warnings and add producer test

ijuma committed:
Introduce FetchRequestTest

ijuma committed:
Preserve behaviour of fetch request version 2 and add test to verify

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2063-bound-fetch-response-size

* apache/trunk:
  KAFKA-4093; Cluster Id (KIP-78)
  KAFKA-4173; SchemaProjector should successfully project missing Struct field when target field is optional
  KAFKA-4183; Corrected Kafka Connect's JSON Converter to properly convert from null to logical values
  KAFKA-3776: Unify store and downstream caching in streams

nepal committed:
Merge pull request #5 from ijuma/kafka-2063-bound-fetch-response-size

KAFKA-2063: Preserve behaviour of fetch requests version 2 and below and add tests

ijuma commented:
This PR is complete, so @nepal, can you please remove the WIP from the PR title and update the PR description to be up to date? The PR title and description became the commit message of the squashed commit that gets merged.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -25,15 +25,12 @@ import kafka.api._
 import kafka.common._
 import kafka.controller.KafkaController
-import kafka.log.{LogAppendInfo, LogManager}
+import kafka.log.{FileMessageSet, LogAppendInfo, LogManager}
```
unused import FileMessageSet

junrao commented on core/src/main/scala/kafka/server/FetchDataInfo.scala:
```diff
@@ -19,4 +19,5 @@ package kafka.server
 
-case class FetchDataInfo(fetchOffsetMetadata: LogOffsetMetadata, messageSet: MessageSet)
+case class FetchDataInfo(fetchOffsetMetadata: LogOffsetMetadata, messageSet: MessageSet,
+                         messageSetIncomplete: Boolean = false)
```
To be clearer, messageSetIncomplete probably should be firstMessageSetIncomplete?

ijuma replied:
Agreed, this also came up in a conversation between Jason and myself and I was thinking of renaming it along the lines you suggest. :)

junrao commented:
Thanks for the latest patch. Made a pass on the server side changes. They look good to me. Just a couple of minor comments, both can be addressed in a followup patch if needed.

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -135,28 +138,21 @@ class ReplicaFetcherThread(name: String,
     }
   }
 
-  def warnIfMessageOversized(messageSet: ByteBufferMessageSet, topicAndPartition: TopicAndPartition): Unit = {
```
This warning may still be useful during rolling upgrade?

ijuma replied:
I thought that it would fix itself once the rolling upgrade is over, but it's probably safer to keep it for now (as you said). Will add it back and add a comment that it only applies to older brokers.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/internals/PartitionStates.java:
```diff
@@ -0,0 +1,171 @@
+ * This class is a useful building block for doing fetch requests where topic partitions have to be rotated via
+ * round-robin to ensure fairness and some level of determinism given the existence of a limit on the fetch response
+ * size. Because the serialization of fetch requests is more efficient if all partitions for the same topic are grouped
+ * together, we do such grouping in the method `set`.
```
For followup: I think it's worth elaborating a little that this heuristic can diverge a little over time from the ideal as partitions are fetched in different orders and partition leadership changes.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -59,29 +64,79 @@ public PartitionData(long offset, int maxBytes) {
+    /**
+     * Create a replica fetch request for the current version.
+     */
+    public static FetchRequest fromReplica(int replicaId, int maxWait, int minBytes, int maxBytes,
```
Static factory methods are totally the way to go for these objects. Having a name makes it much less error-prone and keeps us from having to jump through hoops to keep argument lists different.

ijuma replied:
Are OK with the other constructors remaining as they are for now?

hachikuji replied:
Yeah, I'm fine with it. I was just noting that this is a pattern I'd like to see more of.

hachikuji commented on core/src/main/scala/kafka/api/FetchRequest.scala:
```diff
@@ -56,8 +62,22 @@ object FetchRequest {
   }
+
+  def shuffle(requestInfo: Seq[(TopicAndPartition, PartitionFetchInfo)]): Seq[(TopicAndPartition, PartitionFetchInfo)] =
+    random.shuffle(requestInfo)
```
Do you think it would be a little safer to shuffle by topic and by partition separately to keep the fetch request size consistent with older versions?

ijuma replied:
That's a good idea. I'll do this in a follow-up and include some tests.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -508,78 +512,102 @@ class ReplicaManager(val config: KafkaConfig,
+      } catch {
+        // NOTE: Failed fetch requests metric is not incremented for known exceptions since it
+        // is supposed to indicate un-expected failure of a broker in handling a fetch request
+        case utpe: UnknownTopicOrPartitionException =>
```
minor: Would it be worthwhile to combine some of these cases since they are all handled the same?

hachikuji commented:
@nepal @ijuma Fantastic work on this KIP! I left a couple minor comments which can be addressed in a followup (if needed). Overall, LGTM.

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2063-bound-fetch-response-size

* apache/trunk:
  KAFKA-3492; Secure quotas for authenticated users

ijuma committed:
Rename `messageSetIncomplete` to `firstMessageSetIncomplete`

As per Jun's suggestion.

ijuma committed:
Group exception clauses since the result is the same

As per Jason's suggestion

ijuma committed:
Mention that the way we group topics and partitions deviates from the optimal over time

ijuma committed:
Restore replication warning due to oversized messages if fetch request <= 2

Suggested by Jun

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2063-bound-fetch-response-size

* apache/trunk:
  MINOR: Update the README.md to include a note about GRADLE_USER_HOME
  KAFKA-4157; Transient system test failure in replica_verification_test.test_replica_lags
  HOTFIX: changed quickstart donwload from 0.10.0.0 to 0.10.0.1
  HOTFIX: Increase timeout for bounce test

ijuma committed:
Improve protocol description for fetch request version 3

nepal committed:
Merge pull request #6 from ijuma/kafka-2063-bound-fetch-response-size

KAFKA-2063: Merge conflicts with trunk and address review feedback

ijuma commented:
One failure in the system test run:

test_id:    2016-09-17--001.kafkatest.tests.connect.connect_distributed_test.ConnectDistributedTest.test_bounce.clean=True
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/549/console

It looks unrelated as we had the same failure in trunk recently (http://confluent-kafka-system-test-results.s3-us-west-2.amazonaws.com/2016-09-17--001.1474133882--apache--trunk--ecc1fb1/report.html).

The Jenkins PR builder timed out after 180 minutes while running a Streams test, but all client and core tests passed (I've restarted the build). The tests passed locally.
